src/Microsoft.AspNetCore.Server.IntegrationTesting/TestMatrix.cs

```diff
@@ -0,0 +1,266 @@
+        /// </summary>
+        /// <returns></returns>
+        public TestMatrix WithAncmV2InProcess() => WithAncmVersions(AncmVersion.AspNetCoreModuleV2).WithHostingModels(HostingModel.InProcess);
+
+        public TestList Build()
+        {
+            var data = new TestList();
+            if (!Servers.Any())
+            {
+                // Params error, a server is required. This will cause an xunit error.
+                return data;
+            }
+
+            // TFMs. If not set then use the one from the current app
+            if (!Tfms.Any())
+            {
+                var tfmAttribute = Assembly.GetCallingAssembly().GetCustomAttribute<TargetFrameworkAttribute>();
+                if (tfmAttribute != null && !string.IsNullOrEmpty(tfmAttribute.FrameworkName))
+                {
+                    switch (tfmAttribute.FrameworkName)
+                    {
+                        case ".NETFramework,Version=v4.6.1":
+                            Tfms.Add(Tfm.Net461);
+                            break;
+                        case ".NETCoreApp,Version=v2.0":
+                            Tfms.Add(Tfm.NetCoreApp20);
+                            break;
+                        case ".NETCoreApp,Version=v2.1":
+                            Tfms.Add(Tfm.NetCoreApp21);
+                            break;
+                        case ".NETCoreApp,Version=v2.2":
+                            Tfms.Add(Tfm.NetCoreApp22);
+                            break;
+                    }
+                }
+
+                if (!Tfms.Any())
+                {
+                    throw new ArgumentException("No TFMs was provided and one could be detected from the caller, specify a TFM.");
+                }
+            }
+
+            if (!Architectures.Any())
+            {
+                switch (RuntimeInformation.OSArchitecture)
+                {
+                    case Architecture.X86:
+                        Architectures.Add(RuntimeArchitecture.x86);
+                        break;
+                    case Architecture.X64:
+                        Architectures.Add(RuntimeArchitecture.x64);
+                        break;
+                    default:
+                        throw new ArgumentException(RuntimeInformation.OSArchitecture.ToString());
+                }
+            }
+
+            if (!ApplicationTypes.Any())
+            {
+                ApplicationTypes.Add(ApplicationType.Portable);
+            }
+
+            foreach (var server in Servers)
+            {
+                if (!ServerIsSupportedOnThisOS(server))
+                {
+                    continue;
+                }
+
+                foreach (var tfm in Tfms)
+                {
+                    if (!TfmIsSupportedOnThisOS(tfm))
+                    {
+                        continue;
+                    }
+
+                    foreach (var t in ApplicationTypes)
+                    {
+                        var type = t;
```

Comment:
Why?

Reply:
Otherwise it won't let me overwrite t inside the loop. foreach params are readonly.